<a href="https://colab.research.google.com/github/GrichinNikita2019/yandex_praktikum_da/blob/main/books_app_sql/books_app_sql.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Анализ информации о книгах, издательствах, авторах, а также о пользовательских обзоров книг

**Описание проекта:**

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

**Цель проекта:** проанализировать информацию о книгах, издательствах, авторах, а также о пользовательских обзорах книг


**Задачи проекта**:

1. Посчитайте, сколько книг вышло после 1 января 2000 года.
2. Для каждой книги посчитайте количество обзоров и среднюю оценку.
3. Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры.
4. Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками.
5. Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

**План работы:**
1. Шаг 1. Исследуйте таблицы — выведите первые строки,
2. Шаг 2. Сделайте по одному SQL-запросу для решения каждого задания,
3. Шаг 3. Выведите результат каждого запроса в тетрадке,
4. Шаг 4. Опишите выводы по каждой из решённых задач.

**Описание данных (структура таблицы books):**

* book_id — *идентификатор книги*
* author_id — *идентификатор автора*
* title — *название книги*
* num_pages — *количество страниц*
* publication_date — *дата публикации книги*
* publisher_id — *идентификатор издателя*

**Описание данных (структура таблицы authors):**

* author_id — *идентификатор автора*
* author — *имя автора*

**Описание данных (структура таблицы publishers):**

* publisher_id — *идентификатор издательства*
* publisher — *название издательства*

**Описание данных (структура таблицы ratings):**

* rating_id — *идентификатор оценки*
* book_id — *идентификатор книги*
* username — *имя пользователя, оставившего оценку*
* rating — *оценка книги*

**Описание данных (структура таблицы reviews):**

* review_id — *идентификатор обзора*
* book_id — *идентификатор книги*
* username — *имя пользователя, написавшего обзор*
* text — *текст обзора*

## Импорт библиотек, настройка отображения данных и подключение к базе данных

In [ ]:
import pandas as pd
from sqlalchemy import create_engine

In [ ]:
pd.options.display.float_format = '{:,.2f}'.format
pd.set_option('display.float_format', lambda x: '%.3f' % x)

In [ ]:
db_config = {'user': 'praktikum_student',
 'pwd': 'Sdf4$2;d-d30pp',
 'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
 'port': 6432,
 'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

In [ ]:
def sql_result(query):
  return pd.io.sql.read_sql(query, con = engine)

## Проверка данных

**Задача**:

Исследуйте таблицы — выведите первые строки

Запрос к таблице *books*:

In [ ]:
query = \
'''
select * 
from books
limit 5
'''
sql_result(query)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

Запрос работает корректно.

Запрос к таблице *authors*:

In [ ]:
query = \
'''
select * 
from authors
limit 5
'''
sql_result(query)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

Запрос работает корректно.

Запрос к таблице *publishers*:

In [ ]:
query = \
'''
select * 
from publishers
limit 5
'''
sql_result(query)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

Запрос работает корректно.

Запрос к таблице *ratings*:

In [ ]:
query = \
'''
select * 
from ratings
limit 5
'''
sql_result(query)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

Запрос работает корректно.

Запрос к таблице *reviews*:

In [ ]:
query = \
'''
select * 
from reviews
limit 5
'''
sql_result(query)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

Запрос работает корректно.

## Исследовательский анализ данных

**Задания**:

- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

**Задача 1**:

Посчитайте, сколько книг вышло после 1 января 2000 года

In [ ]:
query = \
'''
select count (book_id) as count_of_books
from books
where publication_date > '2000-01-01'
'''
count_of_books = sql_result(query)
print("С 1 января 200 года вышло {} книг".format(int(count_of_books.iloc[0])))

С 1 января 200 года вышло 819 книг


**Задача 2**:

Для каждой книги посчитайте количество обзоров и среднюю оценку

In [ ]:
query = \
'''
select books.title,
       count (distinct review_id) as count_review,
       avg(rating) AS avg_rating
from books
left join reviews on reviews.book_id = books.book_id
left join ratings on ratings.book_id = books.book_id
group by books.title
order by count(reviews.review_id) desc
'''
print("Количество обзоров и средняя оценка представена в таблице:")
sql_result(query)

Количество обзоров и средняя оценка представена в таблице:


title  count_review  \
0                              Twilight (Twilight  #1)             7   
1                  The Hobbit  or There and Back Again             6   
2                               The Catcher in the Rye             6   
3    Harry Potter and the Prisoner of Azkaban (Harr...             6   
4    Harry Potter and the Chamber of Secrets (Harry...             6   
..                                                 ...           ...   
994  Disney's Beauty and the Beast (A Little Golden...             0   
995                          Essential Tales and Poems             0   
996                               Leonardo's Notebooks             0   
997                            The Natural Way to Draw             0   
998   The Cat in the Hat and Other Dr. Seuss Favorites             0   

     avg_rating  
0         3.663  
1         4.125  
2         3.826  
3         4.415  
4         4.287  
..          ...  
994       4.000  
995       4.000  
996       4.000  
997       3.000  
998       5.000  

[999 rows x 3 columns]

Топ-5 позиций по количеству отзывов занимают книги:
  - Twilight: 7 отзывов
  - The Hobbit or There and Back Again: 6 отзывов
  - The Catcher in the Rye: 6 отзывов
  - Harry Potter and the Prisoner of Azkaban: 6 отзывов
  - Harry Potter and the Chamber of Secrets: 6 отзывов

**Задача 3**:

Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры

In [ ]:
query = \
'''
select publisher,
       count (book_id) as count_of_books
from books
left join publishers on publishers.publisher_id=books.publisher_id
where num_pages > 50
group by publisher
order by count_of_books desc
'''
count_of_books = sql_result(query)
print('''Издательство, которое выпустило наибольшее число книг толще 50 страниц: {}.
Оно выпустило {} книги.'''.format(count_of_books['publisher'][0],
                                  count_of_books['count_of_books'][0]))

Издательство, которое выпустило наибольшее число книг толще 50 страниц: Penguin Books.
Оно выпустило 42 книги.


**Задача 4**:

Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [ ]:
query = \
'''
select author,
       sub.avg_rating as avg_rating
from
(select author,
        avg(rating) AS avg_rating
from authors
left join books on books.author_id = authors.author_id
left join ratings on ratings.book_id = books.book_id
group by author
having count(rating) > 50) AS sub
group by author,
         avg_rating
order by avg_rating desc
'''

author = sql_result(query)
print('''Автор с самой высокой средней оценкой книг: {}.
Его средняя оценка {}.'''.format(author['author'][0],
                                 author['avg_rating'][0]))

Автор с самой высокой средней оценкой книг: J.K. Rowling/Mary GrandPré.
Его средняя оценка 4.288461538461538.


**Задача 5**:

Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [ ]:
query = \
'''
select avg(review_cnt) as avg_review_cnt
from
(select count(review_id) as review_cnt
from reviews
where username in
(select username
from ratings
group by username
having count(ratings) > 50)
group by username) as sub_review_cnt 
'''
avg_review_cnt = sql_result(query)
print('''Cреднее количество обзоров от пользователей, которые поставили больше 50 оценок: {}.'''.format(avg_review_cnt['avg_review_cnt'][0]))

Cреднее количество обзоров от пользователей, которые поставили больше 50 оценок: 24.333333333333332.


## Основные выводы

1. С 1 января 200 года вышло 819 книг
2. Топ-5 позиций по количеству отзывов занимают книги:
  - Twilight: 7 отзывов
  - The Hobbit or There and Back Again: 6 отзывов
  - The Catcher in the Rye: 6 отзывов
  - Harry Potter and the Prisoner of Azkaban: 6 отзывов
  - Harry Potter and the Chamber of Secrets: 6 отзывов
3. Издательство, которое выпустило наибольшее число книг толще 50 страниц: Penguin Books. Оно выпустило 42 книги.
4. Автор с самой высокой средней оценкой книг: J.K. Rowling/Mary GrandPré.
Его средняя оценка 4.283844038868471.
5. Cреднее количество обзоров от пользователей, которые поставили больше 50 оценок: 24.333333333333332.